### PR TITLE
Add wallets management functions

### PR DESCRIPTION
**Pull request by promag**
This is a small step towards dynamic wallet load/unload. The wallets *registry* `vpwallets` is used in several places. With these new functions all `vpwallets` usage are removed and `vpwallets` is now a static variable (no external linkage).

The typedef `CWalletRef` is also removed as it is narrowly used.

### CODE REVIEW

**Comment by MarcoFalke** [comment]
Make sure to align this with the discussion in https://github.com/bitcoin/bitcoin/pull/10740#issuecomment-380950109

File: src/wallet/wallet.cpp
Comment (Empact) [inline src/wallet/wallet.cpp]: Do we want to provide guarantees that `GetWallets` returns no `nullptrs`? Could check here for that.

File: src/wallet/wallet.cpp
Comment (Empact) [inline src/wallet/wallet.cpp]: Do we want to provide guarantees that `GetWallets` returns no `nullptrs`? Could check here for that.

**Comment by Empact** [comment]
If we're doing shared ptrs alal #11402, `CWalletRef` would be preferable to `auto` IMO. https://github.com/bitcoin/bitcoin/pull/11402#discussion_r140923964

File: src/wallet/wallet.cpp
Comment (promag) [reply src/wallet/wallet.cpp]: Added assertions in `AddWallet` and `RemoveWallet`. Also added a check to prevent adding the same wallet (`RemoveWallet` already checks if wallet exists).

**Comment by promag** [comment]
@MarcoFalke afaiu yes. Having `vpwallets` centralised helps in adding concurrency and eventually switching to shared pointers. I've also kept these as global functions as I'm not sure if these can belong to `WalletInitInterface` (imo no).

@Empact thanks for the review. I'm not saying we should use `auto` (it's not used at the moment). I'm not sure if consensus is to use `CWalletRef` instead of `std::shared_ptr<CWallet>` (when that happens). I can remove that commit if it's controversial.

**Review by jnewbery — COMMENTED** [review]
Tested ACK 689de7c08605643ebe48096c59e6810fbe3dc605.

Nice cleanup, one comment inline. Please squash final commit.

File: src/wallet/wallet.h
Comment (promag) [reply src/wallet/wallet.h]: But it also returns true for 2 or more.

Comment (promag) [reply src/wallet/wallet.h]: HaveWallets?

Comment (jnewbery) [reply src/wallet/wallet.h]: no need to overcomplicate it. `HasWallets()` is fine. I was just pointing out that the name could be a little bit more precise :slightly_smiling_face:

**Comment by promag** [comment]
@jnewbery squashed.

File: src/wallet/rpcwallet.cpp
Comment (promag) [inline src/wallet/rpcwallet.cpp]: Note to reviewers, I've switched these instructions. Inverted condition to throw the error.

**Comment by Empact** [comment]
could you clang-format? Some whitespace inconsistencies.

**Comment by promag** [comment]
@Empact done, sorry.

**Comment by promag** [comment]
Latests changes are:
 - fix include order and whitespace
 - removed `auto` usage
 - added return `bool` to `AddWallet` and `RemoveWallet`.

**Comment by Empact** [comment]
utACK 3c058fd

File: src/wallet/wallet.cpp
Comment (kallewoof) [inline src/wallet/wallet.cpp]: μNit: Maybe you could add an
```C++
inline bool HasWallet(CWallet* wallet)
{
    assert(wallet);
    return std::find(vpwallets.begin(), vpwallets.end(), wallet) != vpwallets.end();
}
```
and use that here and below (also removing the assert in both places since it's done inline here).

**Review by kallewoof — APPROVED** [review]
utACK 3c058fdcc8a71d17296973cb7f09e44a310df22e

Nice clean-up. Agree about removing `CWalletRef` as it is mostly unused (and ambiguous -- I assumed it was a `std::shared_ptr<CWallet>` like `CTransactionRef`).

File: src/wallet/wallet.cpp
Comment (promag) [reply src/wallet/wallet.cpp]: Your suggestion "conflicts" with adding a critical section here. If I do as you say then the lock will be recursive, something we want to avoid AFAICK.

**Comment by promag** [comment]
Thanks for the review @kallewoof. Like @MarcoFalke said, this is a small piece of a large feature and I'd like to keep the PR scope on focus.

File: src/wallet/wallet.cpp
Comment (kallewoof) [reply src/wallet/wallet.cpp]: That's a pity, but no big deal.

File: src/wallet/wallet.h
Comment (promag) [inline src/wallet/wallet.h]: I can remove these whitespace change if that is controversial.

**Comment by practicalswift** [comment]
Concept ACK

**Comment by MarcoFalke** [comment]
utACK 3c058fdcc8a71d17296973cb7f09e44a310df22e

File: src/wallet/wallet.h
Comment (jnewbery) [reply src/wallet/wallet.h]: no - please don't invalidate reviewer ACKs unnecessarily

**Comment by laanwj** [comment]
ACK on adding wallet management functions.

> The typedef CWalletRef is also removed as it is narrowly used.

> If we're doing shared ptrs alal #11402, CWalletRef would be preferable to auto IMO. #11402 (comment)

I think this part still needs discussion. `CWalletRef` was introduced very recently as a generalized handle for wallets, to be able to turn it into some kind of smart pointer later. Hardcoding `CWallet*` makes complicates that.

**Comment by promag** [comment]
> I think this part still needs discussion. CWalletRef was introduced very recently

Introduced in #8694, ping @luke-jr.

> Hardcoding CWallet* complicates that.

Switching to shared pointers is not a matter of just changing the typedef.

Anyway, if the commit is controversial I prefer to remove it, request re-ACK and defer this discussion, it's really out of scope here.

**Comment by MarcoFalke** [comment]
It has been used in 12 instances or so, so not something to have a huge discussion about, imo.

**Comment by luke-jr** [comment]
CWalletRef shouldn't be removed. It was specifically added to eventually become a shared_ptr when we have dynamic wallet loading/unloading.

**Comment by jnewbery** [comment]
> CWalletRef shouldn't be removed. It was specifically added to eventually become a shared_ptr when we have dynamic wallet loading/unloading.

This PR doesn't make changing to shared pointers any more or less difficult.

#10740 builds on this. Like @MarcoFalke says, this is only used in a very few places, so let's not hold up this PR for this.

**Comment by laanwj** [comment]
>  it's really out of scope here.

Yes, that's how I felt about that commit in the first place. Whether to have CWalletRef or not is not part of "Add wallets management functions". 
Anyhow, I'll merge this is if there is broad agreement.

**Pull request merged**
